### DISCUSSION TITLE
I have two questions about alpha blending.

### DISCUSSION THREAD

**Discussion by tugrul** [ImageMagick/ImageMagick#2726] [Digital Image Processing] [answered]
I have two composite images. First logo has absolute black background. Second logo has absolute white background. 

First question: How can I recover original alpha logo from the two composite samples? I discovered this plugin on paint.net forums and it makes that thing pretty close but not exactly same with original state of alpha image because it makes one way determination.

Plugin source: https://gist.github.com/tugrul/ab2f768cfa6b4218dada9deb9e31e9ef

Original alpha logo: 
![alpha_50](https://user-images.githubusercontent.com/163442/96191963-ceb81100-0f4d-11eb-9311-ffcd962c83c4.png)

Black background composite:
![black_50](https://user-images.githubusercontent.com/163442/96191995-daa3d300-0f4d-11eb-9a5d-bec07d748908.png)

White background composite:
![white_50](https://user-images.githubusercontent.com/163442/96192077-f4ddb100-0f4d-11eb-9f1e-b7b116cde65c.png)

Recovered alpha logo from black background composite:
![black_50_unblend](https://user-images.githubusercontent.com/163442/96192123-058e2700-0f4e-11eb-905d-31045edb27c9.png)

Recovered alpha logo from white background composite:
![white_50_unblend](https://user-images.githubusercontent.com/163442/96192138-1048bc00-0f4e-11eb-85ab-20f0bb9289e6.png)

Second question: How can I unblend original alpha logo (it not has matte pixels and all pixels have opacity value) from composite photo (not solid background)?

**Comment by snibgo** [comment]
To solve this, we need to know how the "Black background composite" and "White background composite" were made.

If it is a simple "-compose Over -composite" of the logo image over a black (or white) square, we can use methods shown in [Watermarks](http://im.snibgo.com/watermark.htm). However, I show only how to _guess_ the watermark from an image with and without a watermark. (It is a _guess_ because there are an infinite number of possibilities.) The maths could be developed further to use your two images: watermark on black, and watermark on white.

**Comment by snibgo** [comment, marked as answer]
Okay, I've done the maths, and translated that to IM code. b0.png is the watermark on black, and w0.png is the watermark on white. Windows BAT syntax:
```
magick ^
  -define compose:clamp=off ^
  ( b0.png ^
    ( w0.png +write mpr:D1 ) ^
    -compose MinusSrc -composite ^
    -evaluate Add 100%% +write mpr:Sa ^
  ) ^
  ( mpr:D1 mpr:Sa ^
    -compose Plus -composite ^
    -evaluate Subtract 100%% ^
    mpr:Sa ^
    -compose DivideSrc -composite ^
  ) ^
  +swap ^
  -compose CopyAlpha -composite ^
  x.png
```
x.png is the calculated watermark:
![x](https://user-images.githubusercontent.com/33812368/96204176-8f90bc80-0f5b-11eb-9cea-22f5f1847421.png)

**Comment by tugrul** [comment]
@snibgo It works brilliant. I merged decomposed alpha image to different watermarked image that has different background color (0x304358) and pixels' values are very close to same (99%). I shared my test below.

![compose](https://user-images.githubusercontent.com/163442/96255138-175be280-0fbf-11eb-8336-b8ab566f9399.png)

Edit: I also tested with your dewatermark technique and it worked perfectly even background is complex.

**Comment by tugrul** [comment]
@snibgo I encountered an exception while I trying de-watermark technique. It makes pixels' channels value 255 if they are 0 and the pixels look like burning. This is not related to alpha source because this occurs even alpha pixels are rgba(0,0,0,0).

Original: 
![original](https://user-images.githubusercontent.com/163442/96288389-04abd280-0fec-11eb-9743-978d3581481f.png)

Recovered:
![recovered](https://user-images.githubusercontent.com/163442/96288400-0b3a4a00-0fec-11eb-98f6-fc77daa7dd5c.png)

**Comment by snibgo** [comment]
Please show your command, and input images.

**Comment by tugrul** [comment]
@snibgo you can use the images on my first post.

```
magick ^
  puf_black.png ^
  ( puf_alpha.png ^
    ( +clone -alpha extract +write mpr:WEX ) ^
    -alpha off ^
    -compose Multiply -composite -negate ^
  ) ^
  -gravity center ^
  -compose ModulusAdd -composite ^
  ( mpr:WEX -negate ) ^
  -alpha off ^
  -compose DivideSrc -composite ^
  puf_black_out.png
```

**Comment by snibgo** [comment]
The command I show at [Watermarks: De-watermark](http://im.snibgo.com/watermark.htm#dewm) can't cope with black inputs. Change that command that to:
```
%IM%convert ^
  wm_toes_wm.png ^
  ( wm_mark.png ^
    ( +clone -alpha extract +write mpr:WEX ) ^
    -alpha off ^
    -compose Multiply -composite ^
  ) ^
  -gravity center ^
  -compose MinusSrc -composite ^
  ( mpr:WEX -negate ) ^
  -alpha off ^
  -compose DivideSrc -composite ^
  w.png
```
(I have removed a "-negate", and changed "AddModulus" to "MinusSrc".)

Thanks for discovering this bug.

**Comment by tugrul** [comment]
@snibgo thank you so much. The result is further than I expected. 

There is a few subpixel distortion on solid backgrounds when I look close to the pixels but I think masked blur can solve this 1% loose.

**Comment by snibgo** [comment]
Your logo-on-black and logo-on-white images are 8 bits/channel/pixel. So values have been rounded to integers in the range 0 to 255, so about 0.5% precision. Calculating the watermark uses those results, so will be limited to that precision, and is also rounded to 8-bit integers. Completing the round-trip to a solid colour has the same problem.

For better precision, insert "-depth 16" before writing all outputs. Better still, use "-define quantum:format=floating-point -depth 32" and write to tiff files, so we have better precision floating-point.
